File: Britefury/Tests/BritefuryJ/Parser/ParserTestCase.py
##-*************************
##-* This program is free software; you can use it, redistribute it and/or modify it
##-* under the terms of the GNU General Public License version 2 as published by the
##-* Free Software Foundation. The full text of the GNU General Public License
##-* version 2 can be found in the file named 'COPYING' that accompanies this
##-* program. This source code is (C)copyright Geoffrey French 1999-2008.
##-*************************
from BritefuryJ.DocModel import DMNode
from BritefuryJ.DocModel import DMIOReader, DMIOWriter

import string
import unittest
import cStringIO

import re

_whitespaceRegex = '[ ]*'

def _deepList(xs):
	if isinstance( xs, java.util.List )  or  isinstance( xs, list ):
		return [ _deepList( x )   for x in xs ]
	else:
		return xs

class ParserTestCase (unittest.TestCase):
	__junk_regex__ = _whitespaceRegex

	def __init__(self, *args, **kwargs):
		super( ParserTestCase, self ).__init__( *args, **kwargs )




	def _parseStringTest(self, parser, input, expected, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		result = parser.traceParseStringChars( input, ignoreChars )

		if not result.isValid():
			print 'PARSE FAILURE while parsing %s, stopped at %d: %s'  %  ( input, result.end, input[:result.end] )
			print 'EXPECTED:'
			print expected
			self._onError( result )
		self.assert_( result.isValid() )

		value = DMNode.coerce( result.value )
		expected = DMNode.coerce( expected )

		if result.end != len( input ):
			print 'INCOMPLETE PARSE while parsing', input
			print 'Parsed %d/%d characters'  %  ( result.end, len( input ) )
			print input[:result.end]
			print 'EXPECTED:'
			print expected
			print 'RESULT:'
			print value
			self._onError( result )

		bSame = value == expected
		if not bSame:
			print 'While parsing', input
			print 'EXPECTED:'
			print expected
			print ''
			print 'RESULT:'
			print value
			self._onError( result )
		self.assert_( bSame )



	def _parseStringTestSX(self, parser, input, expectedSX, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		result = parser.traceParseStringChars( input, ignoreChars )

		expected = DMIOReader.readFromString( expectedSX )

		if not result.isValid():
			print 'PARSE FAILURE while parsing', input
			print 'EXPECTED:'
			print expectedSX
			self._onError( result )
		self.assert_( result.isValid() )

		value = DMNode.coerce( result.value )
		expected = DMNode.coerce( expected )

		if result.end != len( input ):
			print 'INCOMPLETE PARSE while parsing', input
			print 'Parsed %d/%d characters'  %  ( result.end, len( input ) )
			print input[:result.end]
			print 'EXPECTED:'
			print expectedSX
			print 'RESULT:'
			print DMIOWriter.writeAsString( value )
			self._onError( result )

		bSame = value == expected
		if not bSame:
			print 'While parsing', input
			print 'EXPECTED:'
			print expectedSX
			print ''
			print 'RESULT:'
			print DMIOWriter.writeAsString( value )
			self._onError( result )
		self.assert_( bSame )


	def _parseStringFailTest(self, parser, input, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		result = parser.traceParseStringChars( input, ignoreChars )

		if result.isValid()   and   result.end == len( input ):
			print 'While parsing', input
			print 'EXPECTED:'
			print '<fail>'
			print ''
			print 'RESULT:'
			print result.value
			print 'consumed %d/%d chars'  %  ( result.end, len( input ) )
			self._onError( result )
		self.assert_( not result.isValid()  or  result.end != len( input ) )




	def _parseNodeTest(self, parser, input, expected, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		result = parser.traceParseNode( input, ignoreChars )

		if not result.isValid():
			print 'PARSE FAILURE while parsing %s'  %  ( input, )
			print 'EXPECTED:'
			print expected
			self._onError( result )
		self.assert_( result.isValid() )

		value = DMNode.coerce( result.value )
		expected = DMNode.coerce( expected )

		bSame = value == expected
		if not bSame:
			print 'While parsing', input
			print 'EXPECTED:'
			print expected
			print ''
			print 'RESULT:'
			print value
			self._onError( result )
		self.assert_( bSame )



	def _parseNodeTestSX(self, parser, inputXS, expectedSX, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		input = DMIOReader.readFromString( inputXS )
		expected = DMIOReader.readFromString( expectedSX )

		result = parser.traceParseNode( input, ignoreChars )

		if not result.isValid():
			print 'PARSE FAILURE while parsing', input
			print 'EXPECTED:'
			print expectedSX
			self._onError( result )
		self.assert_( result.isValid() )

		value = DMNode.coerce( result.value )
		expected = DMNode.coerce( expected )

		bSame = value == expected
		if not bSame:
			print 'While parsing', input
			print 'EXPECTED:'
			print expectedSX
			print ''
			print 'RESULT:'
			print DMIOWriter.writeAsString( value )
			self._onError( result )
		self.assert_( bSame )


	def _parseNodeFailTest(self, parser, input, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		result = parser.traceParseNode( input, ignoreChars )

		if result.isValid()   and   result.end == len( input ):
			print 'While parsing', input
			print 'EXPECTED:'
			print '<fail>'
			print ''
			print 'RESULT:'
			print result.value
			print 'consumed %d/%d chars'  %  ( result.end, len( input ) )
			self._onError( result )
		self.assert_( not result.isValid()  or  result.end != len( input ) )


	def _parseNodeFailTestSX(self, parser, inputSX, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		input = DMIOReader.readFromString( inputXS )

		result = parser.traceParseNode( input, ignoreChars )

		if result.isValid()   and   result.end == len( input ):
			print 'While parsing', input
			print 'EXPECTED:'
			print '<fail>'
			print ''
			print 'RESULT:'
			print result.value
			print 'consumed %d/%d chars'  %  ( result.end, len( input ) )
			self._onError( result )
		self.assert_( not result.isValid()  or  result.end != len( input ) )





	def _parseListTest(self, parser, input, expected, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		result = parser.traceParseListItems( input, ignoreChars )

		if not result.isValid():
			print 'PARSE FAILURE while parsing %s, stopped at %d: %s'  %  ( input, result.end, input[:result.end] )
			print 'EXPECTED:'
			print expected
			self._onError( result )
		self.assert_( result.isValid() )

		value = DMNode.coerce( result.value )
		expected = DMNode.coerce( expected )

		if result.end != len( input ):
			print 'INCOMPLETE PARSE while parsing', input
			print 'Parsed %d/%d items'  %  ( result.end, len( input ) )
			print input[:result.end]
			print 'EXPECTED:'
			print expected
			print 'RESULT:'
			print value
			self._onError( result )

		bSame = value == expected
		if not bSame:
			print 'While parsing', input
			print 'EXPECTED:'
			print expected
			print ''
			print 'RESULT:'
			print value
			self._onError( result )
		self.assert_( bSame )



	def _parseListTestSX(self, parser, inputSX, expectedSX, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		input = DMIOReader.readFromString( inputSX )
		expected = DMIOReader.readFromString( expectedSX )

		result = parser.traceParseListItems( input, ignoreChars )

		if not result.isValid():
			print 'PARSE FAILURE while parsing', input
			print 'EXPECTED:'
			print expectedSX
			self._onError( result )
		self.assert_( result.isValid() )

		value = DMNode.coerce( result.value )
		expected = DMNode.coerce( expected )

		if result.end != len( input ):
			print 'INCOMPLETE PARSE while parsing', input
			print 'Parsed %d/%d items'  %  ( result.end, len( input ) )
			print input[:result.end]
			print 'EXPECTED:'
			print expectedSX
			print 'RESULT:'
			print DMIOWriter.writeAsString( value )
			self._onError( result )

		bSame = value == expected
		if not bSame:
			print 'While parsing', input
			print 'EXPECTED:'
			print expectedSX
			print ''
			print 'RESULT:'
			print DMIOWriter.writeAsString( value )
			self._onError( result )
		self.assert_( bSame )


	def _parseListFailTest(self, parser, input, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		result = parser.traceParseListItems( input, ignoreChars )

		if result.isValid()   and   result.end == len( input ):
			print 'While parsing', input
			print 'EXPECTED:'
			print '<fail>'
			print ''
			print 'RESULT:'
			print result.value
			print 'consumed %d/%d items'  %  ( result.end, len( input ) )
			self._onError( result )
		self.assert_( not result.isValid()  or  result.end != len( input ) )



	def _parseListFailTestSX(self, parser, inputSX, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		input = DMIOReader.readFromString( inputSX )

		result = parser.traceParseListItems( input, ignoreChars )

		if result.isValid()   and   result.end == len( input ):
			print 'While parsing', input
			print 'EXPECTED:'
			print '<fail>'
			print ''
			print 'RESULT:'
			print result.value
			print 'consumed %d/%d items'  %  ( result.end, len( input ) )
			self._onError( result )
		self.assert_( not result.isValid()  or  result.end != len( input ) )







	def _parseRichStringTest(self, parser, input, expected, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		result = parser.traceParseRichStringItems( input, ignoreChars )

		if not result.isValid():
			print 'PARSE FAILURE while parsing %s, stopped at %d: %s'  %  ( input, result.end, input[:result.end] )
			print 'EXPECTED:'
			print expected
			self._onError( result )
		self.assert_( result.isValid() )

		value = DMNode.coerce( result.value )
		expected = DMNode.coerce( expected )

		if result.end != len( input ):
			print 'INCOMPLETE PARSE while parsing', input
			print 'Parsed %d/%d items'  %  ( result.end, len( input ) )
			print input[:result.end]
			print 'EXPECTED:'
			print expected
			print 'RESULT:'
			print value
			self._onError( result )

		bSame = value == expected
		if not bSame:
			print 'While parsing', input
			print 'EXPECTED:'
			print expected
			print ''
			print 'RESULT:'
			print value
			self._onError( result )
		self.assert_( bSame )



	def _parseRichStringTestSX(self, parser, input, expectedSX, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		expected = DMIOReader.readFromString( expectedSX )

		result = parser.traceParseListItems( input, ignoreChars )

		if not result.isValid():
			print 'PARSE FAILURE while parsing', input
			print 'EXPECTED:'
			print expectedSX
			self._onError( result )
		self.assert_( result.isValid() )

		value = DMNode.coerce( result.value )
		expected = DMNode.coerce( expected )

		if result.end != len( input ):
			print 'INCOMPLETE PARSE while parsing', input
			print 'Parsed %d/%d items'  %  ( result.end, len( input ) )
			print input[:result.end]
			print 'EXPECTED:'
			print expectedSX
			print 'RESULT:'
			print DMIOWriter.writeAsString( value )
			self._onError( result )

		bSame = value == expected
		if not bSame:
			print 'While parsing', input
			print 'EXPECTED:'
			print expectedSX
			print ''
			print 'RESULT:'
			print DMIOWriter.writeAsString( value )
			self._onError( result )
		self.assert_( bSame )


	def _parseRichStringFailTest(self, parser, input, ignoreChars=None):
		if ignoreChars is None:
			ignoreChars = self.__junk_regex__
		result = parser.traceParseListItems( input, ignoreChars )
		
		if result.isValid()   and   result.end == len( input ):
			print 'While parsing', input
			print 'EXPECTED:'
			print '<fail>'
			print ''
			print 'RESULT:'
			print result.value
			print 'consumed %d/%d items'  %  ( result.end, len( input ) )
			self._onError( result )
		self.assert_( not result.isValid()  or  result.end != len( input ) )
		
		
		
	def _onError(self, parseResult):
		pass
	

File: java/tests/ParserOld/ParserTestCase.java
//##* This program is free software; you can use it, redistribute it and/or modify it
//##* under the terms of the GNU General Public License version 2 as published by the
//##* Free Software Foundation. The full text of the GNU General Public License
//##* version 2 can be found in the file named 'COPYING' that accompanies this
//##* program. This source code is (C)copyright Geoffrey French 2008.
//##************************
package tests.ParserOld;

import junit.framework.TestCase;
import BritefuryJ.DocModel.DMIOReader;
import BritefuryJ.DocModel.DMModuleResolver;
import BritefuryJ.DocModel.DMIOReader.BadModuleNameException;
import BritefuryJ.DocModel.DMIOReader.ParseErrorException;
import BritefuryJ.DocModel.DMModule.UnknownClassException;
import BritefuryJ.DocModel.DMModuleResolver.CouldNotResolveModuleException;
import BritefuryJ.Parser.ItemStream.ItemStream;
import BritefuryJ.ParserOld.ParseResult;
import BritefuryJ.ParserOld.ParserExpression;

abstract public class ParserTestCase extends TestCase
{
	abstract protected DMModuleResolver getModuleResolver();

	
	
	private Object readExpectedSX(String expectedSX)
	{
		Object expected = null;
		try
		{
			expected = DMIOReader.readFromString( expectedSX, getModuleResolver() );
		}
		catch (ParseErrorException e)
		{
			System.out.println( "Could not parse expected SX" );
			fail();
		}
		catch (BadModuleNameException e)
		{
			System.out.println( "Bad module name - expected" );
			fail();
		}
		catch (UnknownClassException e)
		{
			System.out.println( "Unknown class name - expected" );
			fail();
		}
		catch (CouldNotResolveModuleException e)
		{
			System.out.println( "Could not resolve module - expected" );
			fail();
		}
		return expected;
	}
	
	public void matchTestSX(ParserExpression parser, String input, String expectedSX)
	{
		matchTestSX( parser, input, expectedSX, "[ \t\n]*" );
	}

	public void matchTestSX(ParserExpression parser, String input, String expectedSX, String ignoreCharsRegex)
	{
		Object expected = readExpectedSX( expectedSX );
		matchTest( parser, input, expected, ignoreCharsRegex );
	}
	
	
	public void matchTest(ParserExpression parser, String input, Object expected)
	{
		matchTest( parser, input, expected, "[ \t\n]*" );
	}

	public void matchTest(ParserExpression parser, String input, Object expected, String ignoreCharsRegex)
	{
		ParseResult result = parser.parseString( input, ignoreCharsRegex );
		
		if ( !result.isValid() )
		{
			System.out.println( "PARSE FAILURE while parsing " + input + ", stopped at " + String.valueOf( result.getEnd() ) + ": " + input.substring(  0, result.getEnd() ) );
			System.out.println( "EXPECTED:" );
			System.out.println( expected.toString() );
		}
		assertTrue( result.isValid() );
		
		Object value = result.getValue();
		String valueStr = value != null  ?  value.toString()  :  "<null>";
		String valueClassName = value != null  ?  value.getClass().getName()  :  "<null>";
		String expectedStr = expected != null  ?  expected.toString()  :  "<null>";
		String expectedClassName = expected != null  ?  expected.getClass().getName()  :  "<null>";
		
		if ( result.getEnd() != input.length() )
		{
			System.out.println( "INCOMPLETE PARSE while parsing " + input );
			System.out.println( "Parsed " + String.valueOf( result.getEnd() ) + "/" + String.valueOf( input.length() ) + " characters" );
			System.out.println( "Parsed text " + input.substring( 0, result.getEnd() ) );
			System.out.println( "EXPECTED: (a " + expectedClassName + ")" );
			System.out.println( expectedStr );
			System.out.println( "RESULT: (a " + valueClassName + ")" );
			System.out.println( valueStr );
		}
		assertEquals( result.getEnd(), input.length() );
		
		
		boolean bValuesMatch = true;
		if ( value == null )
		{
			bValuesMatch = expected == null;
		}
		else
		{
			bValuesMatch = value.equals( expected );
		}
		if ( !bValuesMatch )
		{
			System.out.println( "VALUE DIFFERS FROM EXPECTED" );
			System.out.println( "EXPECTED: (a " + expectedClassName + ")" );
			System.out.println( expectedStr );
			System.out.println( "RESULT: (a " + valueClassName + ")" );
			System.out.println( valueStr );
		}
		assertTrue( bValuesMatch );
	}



	
	
	
	public void matchTest(ParserExpression parser, ItemStream input, Object expected)
	{
		matchTest( parser, input, expected, "[ \t\n]*" );
	}

	public void matchTest(ParserExpression parser, ItemStream input, Object expected, String ignoreCharsRegex)
	{
		ParseResult result = parser.parseStream( input, ignoreCharsRegex );
		
		if ( !result.isValid() )
		{
			System.out.println( "PARSE FAILURE while parsing " + input + ", stopped at " + String.valueOf( result.getEnd() ) + ": " + input.subStream( 0, result.getEnd() ).toString() );
			System.out.println( "EXPECTED:" );
			System.out.println( expected.toString() );
		}
		assertTrue( result.isValid() );
		
		Object value = result.getValue();
		String valueStr = value != null  ?  value.toString()  :  "<null>";
		String valueClassName = value != null  ?  value.getClass().getName()  :  "<null>";
		String expectedStr = expected != null  ?  expected.toString()  :  "<null>";
		String expectedClassName = expected != null  ?  expected.getClass().getName()  :  "<null>";
		
		if ( result.getEnd() != input.length() )
		{
			System.out.println( "INCOMPLETE PARSE while parsing " + input );
			System.out.println( "Parsed " + String.valueOf( result.getEnd() ) + "/" + String.valueOf( input.length() ) + " characters" );
			System.out.println( "Parsed text " + input.subStream( 0, result.getEnd() ).toString() );
			System.out.println( "EXPECTED: (a " + expectedClassName + ")" );
			System.out.println( expectedStr );
			System.out.println( "RESULT: (a " + valueClassName + ")" );
			System.out.println( valueStr );
		}
		assertEquals( result.getEnd(), input.length() );
		
		
		boolean bValuesMatch = true;
		if ( value == null )
		{
			bValuesMatch = expected == null;
		}
		else
		{
			bValuesMatch = value.equals( expected );
		}
		if ( !bValuesMatch )
		{
			System.out.println( "VALUE DIFFERS FROM EXPECTED" );
			System.out.println( "EXPECTED: (a " + expectedClassName + ")" );
			System.out.println( expectedStr );
			System.out.println( "RESULT: (a " + valueClassName + ")" );
			System.out.println( valueStr );
		}
		assertTrue( bValuesMatch );
	}
	
	public void matchTestSX(ParserExpression parser, ItemStream input, String expectedSX)
	{
		matchTestSX( parser, input, expectedSX, "[ \t\n]*" );
	}

	public void matchTestSX(ParserExpression parser, ItemStream input, String expectedSX, String ignoreCharsRegex)
	{
		Object expected = readExpectedSX( expectedSX );
		matchTest( parser, input, expected, ignoreCharsRegex );
	}
	
	
	
	public void matchSubTestSX(ParserExpression parser, String input, String expectedSX, int end)
	{
		matchSubTestSX( parser, input, expectedSX, end, "[ \t\n]*" );
	}

	public void matchSubTestSX(ParserExpression parser, String input, String expectedSX, int end, String ignoreCharsRegex)
	{
		Object expected = readExpectedSX( expectedSX );
		matchSubTest( parser, input, expected, end, ignoreCharsRegex );

	}
	
	
	public void matchSubTest(ParserExpression parser, String input, Object expected, int end)
	{
		matchSubTest( parser, input, expected, end, "[ \t\n]*" );
	}
	
	public void matchSubTest(ParserExpression parser, String input, Object expected, int end, String ignoreCharsRegex)
	{
		ParseResult result = parser.parseString( input, ignoreCharsRegex );

		if ( !result.isValid() )
		{
			System.out.println( "PARSE FAILURE while parsing " + input.substring(  0, end ) + ", stopped at " + String.valueOf( result.getEnd() ) + ": " + input.substring(  0, result.getEnd() ) );
			System.out.println( "EXPECTED:" );
			System.out.println( expected.toString() );
		}
		assertTrue( result.isValid() );
		
		Object value = result.getValue();
		String valueStr = value != null  ?  value.toString()  :  "<null>";
		String valueClassName = value != null  ?  value.getClass().getName()  :  "<null>";
		String expectedStr = expected != null  ?  expected.toString()  :  "<null>";
		String expectedClassName = expected != null  ?  expected.getClass().getName()  :  "<null>";
		
		if ( result.getEnd() != end )
		{
			System.out.println( "DID NOT PARSE CORRECT AMOUNT while parsing " + input.substring( 0, end ) );
			System.out.println( "Parsed " + String.valueOf( result.getEnd() ) + "/" + String.valueOf( end ) + " characters" );
			System.out.println( input.substring( 0, result.getEnd() ) );
			System.out.println( "EXPECTED: (a " + expectedClassName + ")" );
			System.out.println( expectedStr );
			System.out.println( "RESULT: (a " + valueClassName + ")" );
			System.out.println( valueStr );
		}
		assertEquals( result.getEnd(), end );
		
		
		boolean bValuesMatch = true;
		if ( value == null )
		{
			bValuesMatch = expected == null;
		}
		else
		{
			bValuesMatch = value.equals( expected );
		}
		if ( !bValuesMatch )
		{
			System.out.println( "VALUE DIFFERS FROM EXPECTED" );
			System.out.println( "EXPECTED: (a " + expectedClassName + ")" );
			System.out.println( expectedStr );
			System.out.println( "RESULT: (a " + valueClassName + ")" );
			System.out.println( valueStr );
		}
		assertTrue( bValuesMatch );
	}



	public void matchSubTest(ParserExpression parser, ItemStream input, Object expected, int end)
	{
		matchSubTest( parser, input, expected, end, "[ \t\n]*" );
	}
	
	public void matchSubTest(ParserExpression parser, ItemStream input, Object expected, int end, String ignoreCharsRegex)
	{
		ParseResult result = parser.parseStream( input, ignoreCharsRegex );

		if ( !result.isValid() )
		{
			System.out.println( "PARSE FAILURE while parsing " + input.subStream(  0, end ) + ", stopped at " + String.valueOf( result.getEnd() ) + ": " + input.subStream(  0, result.getEnd() ) );
			System.out.println( "EXPECTED:" );
			System.out.println( expected.toString() );
		}
		assertTrue( result.isValid() );
		
		Object value = result.getValue();
		String valueStr = value != null  ?  value.toString()  :  "<null>";
		String valueClassName = value != null  ?  value.getClass().getName()  :  "<null>";
		String expectedStr = expected != null  ?  expected.toString()  :  "<null>";
		String expectedClassName = expected != null  ?  expected.getClass().getName()  :  "<null>";
		
		if ( result.getEnd() != end )
		{
			System.out.println( "DID NOT PARSE CORRECT AMOUNT while parsing " + input.subStream( 0, end ) );
			System.out.println( "Parsed " + String.valueOf( result.getEnd() ) + "/" + String.valueOf( end ) + " characters" );
			System.out.println( input.subStream( 0, result.getEnd() ) );
			System.out.println( "EXPECTED: (a " + expectedClassName + ")" );
			System.out.println( expectedStr );
			System.out.println( "RESULT: (a " + valueClassName + ")" );
			System.out.println( valueStr );
		}
		assertEquals( result.getEnd(), end );
		
		
		boolean bValuesMatch = true;
		if ( value == null )
		{
			bValuesMatch = expected == null;
		}
		else
		{
			bValuesMatch = value.equals( expected );
		}
		if ( !bValuesMatch )
		{
			System.out.println( "VALUE DIFFERS FROM EXPECTED" );
			System.out.println( "EXPECTED: (a " + expectedClassName + ")" );
			System.out.println( expectedStr );
			System.out.println( "RESULT: (a " + valueClassName + ")" );
			System.out.println( valueStr );
		}
		assertTrue( bValuesMatch );
	}



	public void matchFailTest(ParserExpression parser, String input)
	{
		matchFailTest( parser, input, "[ \t\n]*" );
	}
	
	public void matchFailTest(ParserExpression parser, String input, String ignoreCharsRegex)
	{
		ParseResult result = parser.parseString( input, ignoreCharsRegex );

		if ( result.isValid() )
		{
			Object value = result.getValue();
			String valueStr = value != null  ?  value.toString()  :  "<null>";
			String valueClassName = value != null  ?  value.getClass().getName()  :  "<null>";

			System.out.println( "FAILURE EXPECTED; GOT SUCCESS" );
			System.out.println( "EXPECTED:" );
			System.out.println( "<fail>" );
			System.out.println( "RESULT: (a " + valueClassName + ")" );
			System.out.println( valueStr );
			System.out.println( "Consumed " + String.valueOf( result.getEnd() ) + "/" + String.valueOf( input.length() ) + " characters" );
		}
		assertFalse( result.isValid() );
	}



	public void matchFailTest(ParserExpression parser, ItemStream input)
	{
		matchFailTest( parser, input, "[ \t\n]*" );
	}
	
	public void matchFailTest(ParserExpression parser, ItemStream input, String ignoreCharsRegex)
	{
		ParseResult result = parser.parseStream( input, ignoreCharsRegex );

		if ( result.isValid() )
		{
			Object value = result.getValue();
			String valueStr = value != null  ?  value.toString()  :  "<null>";
			String valueClassName = value != null  ?  value.getClass().getName()  :  "<null>";

			System.out.println( "FAILURE EXPECTED; GOT SUCCESS" );
			System.out.println( "EXPECTED:" );
			System.out.println( "<fail>" );
			System.out.println( "RESULT: (a " + valueClassName + ")" );
			System.out.println( valueStr );
			System.out.println( "Consumed " + String.valueOf( result.getEnd() ) + "/" + String.valueOf( input.length() ) + " characters" );
		}
		assertFalse( result.isValid() );
	}



	public void matchIncompleteTest(ParserExpression parser, String input)
	{
		matchIncompleteTest( parser, input, "[ \t\n]*" );
	}
	
	public void matchIncompleteTest(ParserExpression parser, String input, String ignoreCharsRegex)
	{
		ParseResult result = parser.parseString( input, ignoreCharsRegex );

		if ( !result.isValid() )
		{
			System.out.println( "PARSE FAILURE while parsing " + input + ", stopped at " + String.valueOf( result.getEnd() ) + ": " + input.substring(  0, result.getEnd() ) );
			System.out.println( "EXPECTED INCOMPLETE PARSE:" );
		}
		assertTrue( result.isValid() );
		
		Object value = result.getValue();
		String valueStr = value != null  ?  value.toString()  :  "<null>";
		String valueClassName = value != null  ?  value.getClass().getName()  :  "<null>";
		
		if ( result.getEnd() == input.length() )
		{
			System.out.println( "EXPECTED INCOMPLETE PARSE while parsing " + input );
			System.out.println( "RESULT: (a " + valueClassName + ")" );
			System.out.println( valueStr );
		}
		assertTrue( result.getEnd() != input.length() );
	}




	public void matchIncompleteTest(ParserExpression parser, ItemStream input)
	{
		matchIncompleteTest( parser, input, "[ \t\n]*" );
	}
	
	public void matchIncompleteTest(ParserExpression parser, ItemStream input, String ignoreCharsRegex)
	{
		ParseResult result = parser.parseStream( input, ignoreCharsRegex );

		if ( !result.isValid() )
		{
			System.out.println( "PARSE FAILURE while parsing " + input + ", stopped at " + String.valueOf( result.getEnd() ) + ": " + input.subStream(  0, result.getEnd() ) );
			System.out.println( "EXPECTED INCOMPLETE PARSE:" );
		}
		assertTrue( result.isValid() );
		
		Object value = result.getValue();
		String valueStr = value != null  ?  value.toString()  :  "<null>";
		String valueClassName = value != null  ?  value.getClass().getName()  :  "<null>";
		
		if ( result.getEnd() == input.length() )
		{
			System.out.println( "EXPECTED INCOMPLETE PARSE while parsing " + input );
			System.out.println( "RESULT: (a " + valueClassName + ")" );
			System.out.println( valueStr );
		}
		assertTrue( result.getEnd() != input.length() );
	}
}
